### PR TITLE
docs: Add JOSS paper

### LINKED ISSUE #502: Write JOSS Submission
# Description

Given conversations with @labarba at SciPy 2019 it seems that pyhf would be a good candidate for a [JOSS paper](https://joss.theoj.org/). Given @cranmer's good experience with this for carl [DOI](https://doi.org/10.21105/joss.00011) it seems if we clean up and finish a few of the docstrings then we would be in good shape to submit for a review (once we finish up the other publication process we have ongoing).

The [carl paper submission review](https://github.com/openjournals/joss-reviews/issues/11) might be a good premeptive checklist for this.

### PR DESCRIPTION
**Pull request by matthewfeickert**
# Description

Resolves #502 

Add the text and references for a JOSS paper submission. The paper is created following the [Submitting a paper to JOSS](https://joss.readthedocs.io/en/latest/submitting.html) guidelines.

**N.B.** This branch does not _need_ to be merged to submit to JOSS. (Though it would probably be nice for posterity)

## Action Item Checklist
- [x] Write paper
- [x] [Check it compiles](https://joss.readthedocs.io/en/latest/submitting.html#checking-that-your-paper-compiles)
- [x] [Submit paper to JOSS](https://joss.readthedocs.io/en/latest/submitting.html#submitting-your-paper)

## [Paper Checklist](https://joss.readthedocs.io/en/latest/submitting.html#what-should-my-paper-contain)

Your paper should include:

- [x] A list of the authors of the software and their affiliations, using the correct format.
- [x] A summary describing the high-level functionality and purpose of the software for a diverse, non-specialist audience.
- [x] A clear Statement of Need that illustrates the research purpose of the software.
- [x] A list of key references, including to other software addressing related needs.
- [x] Mention (if applicable) a representative set of past or ongoing research projects using the software and recent scholarly publications enabled by it.
- [x] Acknowledgement of any financial support.

**Paper compilation preview:** https://whedon.theoj.org/
   - Paper repository address: `https://github.com/scikit-hep/pyhf`
   - Custom branch: `docs/add-JOSS-paper` 

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* Add text and references of JOSS paper
* Check submission to JOSS off of ROADMAP
```

### CODE REVIEW

**Comment by matthewfeickert** [comment]
For further reference, here is the [`carl` JOSS paper](https://joss.theoj.org/papers/10.21105/joss.00011) by @glouppe, @cranmer, and @jgpavez.

**Comment by matthewfeickert** [comment]
@lukasheinrich @kratsg Review and edits to the paper are welcome before I submit it. You can see it built here (but you have to build it each time you push):

**Paper compilation preview:** https://whedon.theoj.org/
   - Paper repository address: `https://github.com/scikit-hep/pyhf`
   - Custom branch: `docs/add-JOSS-paper`

**Comment by matthewfeickert** [comment]
Note that there is the [optional submission step](https://joss.readthedocs.io/en/latest/submitting.html#typical-paper-submission-flow) of

> (Optional) create a metadata file describing your software and include it in your repository. We provide a script that automates the generation of this metadata.

which would be part of [Issue #891](https://github.com/scikit-hep/pyhf/issues/891#issuecomment-638509082).

**Review by lukasheinrich — APPROVED** [review]

**Review by kratsg — APPROVED** [review]

**Comment by matthewfeickert** [comment]
@lukasheinrich @kratsg As you both approved this I'm not going to merge this yet, but I will submit it to JOSS for approval. If we don't have any revisions then I'll merge it.

**Comment by matthewfeickert** [comment]
@cranmer can you review this and give any thoughts and revision suggestions? Also please check your JOSS ID to make sure I copied it right.

File: docs/JOSS/paper.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+## Future work
+
+Future development aims to provide support limit setting through pseudoexperiment generation in the regimes in which asymptotic approximations [@Cowan:2010js] are no longer valid.
+Further improvements to the performance of the library as well as API refinement are also planned.
+
+# Acknowledgements
+
+We would like to thank our fellow developers in the Scikit-HEP community for their continued support and feedback.
+Matthew Feickert has received support to work on `pyhf` provided by NSF cooperative agreement OAC-1836650 (IRIS-HEP) and grant OAC-1450377 (DIANA/HEP).
```
Comment (cranmer) [inline docs/JOSS/paper.md:62]: should add Kyle here too.

File: docs/JOSS/paper.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    orcid: 0000-0002-4048-7584
+    affiliation: 1
+  - name: Matthew Feickert^[Corresponding author.]
+    orcid: 0000-0003-4124-7862
+    affiliation: 2
+  - name: Giordon Stark
+    orcid: 0000-0001-6616-3433
+    affiliation: 3
+  - name: Kyle Cranmer
+    orcid: 0000-0002-5769-7094
```
Comment (cranmer) [inline docs/JOSS/paper.md:21]: confirmed

File: docs/JOSS/paper.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+   index: 3
+ - name: New York University
+   index: 4
+date: 5 October 2020
+bibliography: paper.bib
+---
+
+# Summary
+
+Measurements in High Energy Physics (HEP) rely on determining the compatibility of observed collision events with theoretical predictions.
```
Comment (cranmer) [inline docs/JOSS/paper.md:38]: Suggest rewording as:
"Statistical analysis of High Energy Physics (HEP) data relies on quantifying the compatibility of observed collision events with theoretical predictions."

"Measurements" is kind of vague. We use this for searches, and some might not call that a "measurement".

File: docs/JOSS/paper.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+The relationship between them is often formalised in a statistical model $f(\mathbf{x}|\mathbf{\phi})$ describing the probability of data $\mathbf{x}$ given model parameters $\mathbf{\phi}$.
+Given observed data, the likelihood $\mathcal{L}(\mathbf{\phi})$ then serves as the basis to test hypotheses on the parameters $\mathbf{\phi}$.
```
Comment (cranmer) [inline docs/JOSS/paper.md:40]: suggest: ".... serves as the basis to for inference on the parameters $\mathbf{\phi}$."

File: docs/JOSS/paper.md (hunk 1 above)
Comment (matthewfeickert) [reply docs/JOSS/paper.md]: Done. :+1:

File: docs/JOSS/paper.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+For measurements based on binned data (histograms), the `HistFactory` family of statistical models [@Cranmer:1456844] has been widely used in both Standard Model measurements [@HIGG-2013-02] as well as searches for new physics [@ATLAS-CONF-2018-041].
+`pyhf` is the first pure-Python implementation of the `HistFactory` model specification and implements a declarative, plain-text format for describing `HistFactory`-based likelihoods that is targeted for reinterpretation and long-term preservation in analysis data repositories such as HEPData [@Maguire:2017ypu].
```
Comment (cranmer) [inline docs/JOSS/paper.md:42]: could remove "first" here. Not saying that there is something prior, but not sure it adds much.

**Comment by matthewfeickert** [comment]
@lukasheinrich @kratsg @cranmer Final thoughts and suggestions before submission? @lukasheinrich I think you had one or two comments?

File: docs/JOSS/paper.bib [hunk 6]
```diff
@@ -0,0 +1,134 @@
+@techreport{Cranmer:1456844,
+      author        = "Cranmer, Kyle and Lewis, George and Moneta, Lorenzo and
+                       Shibata, Akira and Verkerke, Wouter",
+      title         = "{HistFactory: A tool for creating statistical models for
+                       use with RooFit and RooStats}",
+      institution   = "New York U.",
+      collaboration = "ROOT Collaboration",
```
Comment (lukasheinrich) [inline docs/JOSS/paper.bib:7]: @cranmer should this be "ROOT collaboration"?

Comment (matthewfeickert) [reply docs/JOSS/paper.bib:7]: FWIW, that's just copy pasted from the CDS entry's BibTeX https://cds.cern.ch/record/1456844/export/hx?ln=en.

**Review by kratsg — APPROVED** [review]

**Comment by matthewfeickert** [comment]
Submitted for review with Eloisa Bentivegna as the suggested topic editor.

**Comment by matthewfeickert** [comment]
I changed the references as requested by @kthyng in https://github.com/openjournals/joss-reviews/issues/2734.

**Comment by matthewfeickert** [comment]
The paper is currently under review at https://github.com/openjournals/joss-reviews/issues/2823

**Comment by matthewfeickert** [comment]
The paper has passed JOSS review! :rocket: [Final requirements for publication include](https://github.com/openjournals/joss-reviews/issues/2823#issuecomment-768135982)

> Upon successful completion of the review, authors will make a tagged release of the software, and deposit a copy of the repository with a data-archiving service such as Zenodo or figshare, get a DOI for the archive, and update the review issue thread with the version number and DOI.

This would most easily be achieved by merging this branch to `master` and then cutting a GitHub release to trigger Zenodo and then altering the Zenodo metadata to make the archive title match the submission title.

@lukasheinrich @kratsg Are you both good with having the paper be part of the codebase? Or would you prefer that we have this branch stay orphaned and cut a specific GitHub / Zenodo release for it?

**Review by lukasheinrich — APPROVED** [review]

**Review by kratsg — APPROVED** [review]
happy for it to just be part of master
